File: middleware/schemas/pipeline.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field


class PipelineStartRequest(BaseModel):
    """POST /api/pipeline/start body."""

    models: list[str] = Field(
        ...,
        description='Models to build: ["PD"], ["LGD"], ["EAD"], or ["PD","LGD","EAD","EL"]',
    )
    db_path: Optional[str] = Field(None, description="Override database path")
    config_overrides: Optional[dict[str, Any]] = Field(
        None, description="Optional agent parameter overrides"
    )


class PipelineStartResponse(BaseModel):
    run_id: str
    sse_url: str


class PipelineStatus(BaseModel):
    run_id: str
    status: str  # running | completed | failed
    current_agent: Optional[str] = None
    completed_agents: list[str] = []
    failed_agent: Optional[str] = None
    progress_pct: float = 0.0
    error: Optional[str] = None


class DatasetInfo(BaseModel):
    db_path: str
    table_name: str
    row_count: int
    column_count: int
    columns: list[str]
    sample_rows: list[dict[str, Any]] = []

File: Frontend/src/types/pipeline.ts
export interface PipelineStartRequest {
  models: string[]
  db_path?: string
  config_overrides?: Record<string, unknown>
}

export interface PipelineStartResponse {
  run_id: string
  sse_url: string
}

export interface DatasetInfo {
  db_path: string
  table_name: string
  row_count: number
  column_count: number
  columns: string[]
  sample_rows: Record<string, unknown>[]
}

export type AgentStatus = 'pending' | 'running' | 'completed' | 'failed'

export interface AgentState {
  name: string
  status: AgentStatus
  duration_s?: number
  metrics: Record<string, number>
}
